File: scala/nth-prime/src/main/scala/NthPrime.scala
object NthPrime {
  lazy val odds: Stream[Int] = Stream.from(3, 2)
  lazy val primes: Stream[Int] = 2 #:: odds.filter(n => odds.takeWhile(i => i * i <= n).forall(n % _ > 0))

  def prime(n: Int) = Some(n).filter(_ > 0).map(n => primes.take(n).last)
}

File: kotlin/nth-prime/src/main/kotlin/NthPrime.kt
object Prime {

    fun nth(n: Int): Int {
        require(n > 0) { "There is no zeroth prime." }
        if (n == 1) return 2

        return generateSequence(3) { it + 2 }
            .filter { f -> (3..f / 2).none { f % it == 0 } }
            .take(n - 1).last()

    }
}
